### PR TITLE
feat: add grain-based data loading

### PR DESCRIPTION
**Pull request by fbanelli**
# PR Description: Grain-based Data Loader Integration

## Summary
This PR introduces a new, high-performance data loading stack based on Google's [Grain](https://github.com/google/grain) library. It replaces several components of the legacy scheduler-based system with a more modular `DataSource` architecture, while maintaining full backward compatibility via adapters.

### 1. New `DataSource` API
We have decoupled file reading from scheduling logic. 
- **Base Components**: Added `FileDataSource` and specific implementations for `.mat`, `.h5`, and `.npy` files.
- **Episodic Logic**: Introduced `EpisodicDataSource`, which handles the interleaving of episode chunks at the data source level, making it compatible with Grain's `IndexSampler`.

### 2. Grain Integration
- **DataLoader**: `synthpix.make` now optionally (and preferably) uses `grain.DataLoader`.
- **Adapters**: `GrainSchedulerAdapter` and `GrainEpisodicAdapter` wrap the Grain iterator to provide a seamless transition for existing codebases that depend on the `get_batch()` and `reset()` interface. **Notably, because of these adapters, not a single line of code in the existing samplers required modification.**

### 3. Performance & Parallelism
- **Multithreading**: Leverages Grain's multi-threaded prefetching, which provides significant speedups over the legacy `PrefetchingFlowFieldScheduler`.
- **Selective Multiprocessing**: Enabled `worker_count` for non-episodic data.
- **Order Preservation**: We explicitly enforce `worker_count=0` (multithreading only) for episodic data to guarantee the strict temporal order required for episodes, as multiprocessing in Grain can lead to out-of-order data.

### 4. Validation & Safety
- Added strict checks in `synthpix.make` to prevent invalid worker configurations.
- Added warnings for non-episodic multiprocessing to ensure users are aware of potential order changes (though deterministic per worker count).

### Correctness
Strict output equality tests confirm that the Grain-based stack produces bit-identical results (flow fields and images) compared to the legacy stack for deterministic seeds.

> [!WARNING]
> The broader testing suite still relies on the legacy scheduler stack. This PR provides the infrastructure for Grain, but a separate issue has been opened to fully transition the codebase and eventually remove the old scheduler.

> [!NOTE]
> Detailed benchmarks can be found in `tests/test_performance_comparison.py`.

## Files Modified
- `src/synthpix/make.py`: Entry point for dataset instantiation.
- `src/synthpix/data_sources/*`: New data source implementations.
- `src/synthpix/data_sources/adapter.py`: Adapters for Grain.
- `tests/test_performance_comparison.py`: Performance and order verification tests.
- `tests/test_scheduler/test_scheduler_prefetch.py`: Updated tests for prefetching.

### CODE REVIEW

File: src/main.py [hunk 1]
```diff
@@ -64,16 +64,22 @@ def visualize_and_save(
     logger.info(f"Saved images for {name} to {output_dir}.")
 
 
-def main(config_path: str, output_dir: str, num_images_to_display: int):
+def main(
+    config_path: str,
+    output_dir: str,
+    num_images_to_display: int,
+    use_grain: bool = False,
```
Comment (antonioterpin) [inline src/main.py]: Should we keep only grain? Or at least make it the default?

File: src/synthpix/data_sources/base.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    def load_file(self, file_path: str) -> dict[str, Any]:
+        """Loads a file and returns the data dictionary.
+
+        Args:
+            file_path: Absolute path to the file.
+
+        Returns:
+            Dictionary containing data (e.g. 'flow_fields', 'images1', etc).
+        """
+        pass  # pragma: no cover
```
Comment (antonioterpin) [inline src/synthpix/data_sources/base.py]: You can just remove this line (no need for pass)

File: src/synthpix/make.py [hunk 3]
```diff
@@ -47,8 +66,29 @@ def get_base_scheduler(name: str) -> BaseFlowFieldScheduler:
...
+        raise ValueError(
+            f"DataSource class {name} not found. Should be one of {list(DATA_SOURCES.keys())}."
+        )
+
+    return DATA_SOURCES[name]
+
+
 def make(
     config: str | dict,
+    use_grain_scheduler: bool = False,
```
Comment (antonioterpin) [inline src/synthpix/make.py]: default (see above)?

File: src/synthpix/make.py [hunk 4]
```diff
@@ -180,27 +227,104 @@ def make(
...
+            raise ValueError(
+                f"worker_count must be 0 when using episodic data (episode_length={episode_length}). "
+                "Using multiple workers (multiprocessing) with EpisodicDataSource breaks data order "
+                "because workers consume the interleaved episode chunks independently."
+            )
+        elif not is_episodic and worker_count > 0:
+            logger.warning(
+                f"Using worker_count={worker_count} with non-episodic data. "
+                "This enables multiprocessing in Grain. Notice that the order of the data is not preserved,"
+                "but still deterministic if you keep the same number of workers."
```
Comment (antonioterpin) [inline src/synthpix/make.py]: Grammar

File: src/synthpix/make.py [hunk 5]
```diff
@@ -180,27 +227,104 @@ def make(
             "output_shape": tuple(dataset_config.get("image_shape", (256, 256))),
         }
 
-    # Initialize the base scheduler
-    scheduler = scheduler_class.from_config(kwargs)
+    # Initialize the scheduler (Legacy or Grain)
+    if use_grain_scheduler:
```
Comment (antonioterpin) [inline src/synthpix/make.py:393]: split in two functions? the if -> create_grain_scheduler, else -> create_legacy_scheduler

I wonder if we should remove alltogether the legacy scheduler.

File: tests/test_data_sources/test_adapter.py [hunk 6]
```diff
@@ -0,0 +1,567 @@
...
+            num_epochs=1,
+        ),
+        operations=[grain.Batch(batch_size=2)],
+    )
+
+    adapter = GrainSchedulerAdapter(loader)
+
+    # First batch: indices 0, 1 -> t=0, t=1
+    batch = adapter.get_batch(2)
+    assert batch.flow_fields.shape == (2, 32, 32, 2)
```
Comment (antonioterpin) [inline tests/test_data_sources/test_adapter.py]: add strings that describe why an assert is failing when it happens

assert condition, string

Comment (antonioterpin) [reply tests/test_data_sources/test_adapter.py]: here and everywhere on the tests

File: tests/test_data_sources/test_adapter.py [hunk 7]
```diff
@@ -0,0 +1,567 @@
...
+                "flow_fields": np.zeros((32, 32, 2)),
+                "images2": "not_an_array",
+                "file": ("f",),
+            }
+        ]
+    )
+    # create new adapter to pick up new iterator
+    adapter = GrainSchedulerAdapter(loader)
+    # Ensure include_images is False to avoid KeyError for missing images1
+    loader._data_source.include_images = False
```
Comment (antonioterpin) [inline tests/test_data_sources/test_adapter.py:563]: Add a comment that says that when include_images is False, if the keys are present are still checked and this is what this test covers

File: tests/test_data_sources/test_mat.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
+"""Tests for MATDataSource."""
```
Comment (antonioterpin) [inline tests/test_data_sources/test_mat.py]: Here and everywhere: I think the docstring of each test should have a more descriptive explanation of what is going to be tested. Otherwise, we may have tons of duplicated tests and also hard to know where to look for when one is missing... We need some structure for this.

File: tests/test_data_sources/test_numpy.py [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+    with pytest.raises(FileNotFoundError, match="Missing images"):
+        NumpyDataSource([str(tmp_path)], include_images=True)
+
+
+def test_numpy_invalid_extension(tmp_path):
+    """Test that ValueError is raised if files do not have .npy extension."""
+    p_txt = tmp_path / "test.txt"
+    p_txt.write_text("dummy")
+
+    # NumpyDataSource expects .npy files if it finds any files that match its internal logic?
```
Comment (antonioterpin) [inline tests/test_data_sources/test_numpy.py]: leftover?

File: tests/test_data_sources/test_numpy.py [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+    
+    # Let's test the specific validation logic by mocking file discovery if needed, 
+    # Or just rely on standard usage. 
+    # If I put a file that matches pattern but somehow isn't .npy? 
+    # The pattern is "flow_*.npy". So it must end with .npy. 
+    # The check `if not all(fp.endswith(".npy")` seems redundant if discovery works by glob?
+    # Unless someone manually passes a list to base class (which isn't shown in __init__ args easily).
+    # Wait, FileDataSource takes dataset_path which is list[str] | str.
+    # If it's a file path?
+    pass
```
[the remark this answers is not in the document]
Comment (antonioterpin) [reply tests/test_data_sources/test_numpy.py]: yes

File: tests/test_data_sources/test_numpy.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+    ds._include_images = True
+
+    with pytest.raises(ValueError, match="Bad filename"):
+        ds.load_file(str(p))
+
+
+def test_numpy_manual_list_validation():
+    """Test validation if file list contains non-npy file (manual injection)."""
+    # FileDataSource init usually scans.
+    # But if we assume scan found something invalid, or we subclass/override?
```
Comment (antonioterpin) [inline tests/test_data_sources/test_numpy.py]: leftover?

File: tests/test_performance_comparison.py [hunk 12]
```diff
@@ -0,0 +1,339 @@
+"""Performance comparison between Legacy and Grain schedulers.
```
Comment (antonioterpin) [inline tests/test_performance_comparison.py]: What are the results? If comparable or grain is faster, let's make it default. We can still keep the legacy mostly for comparisons and tests.

Comment (antonioterpin) [reply tests/test_performance_comparison.py]: Also, these are not only performance comparison, but also correctness. Perhaps we could split in two files for tests, and in the correctness one also stress test the corner cases (remainders etc).

File: src/synthpix/data_sources/hdf5.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        """Loads the dataset from the HDF5 file.
+
+        Args:
+            file_path: Path to the .h5 file.
+
+        Returns:
+            Dictionary containing data (e.g. 'flow_fields', 'images1', etc).
+        """
+        data = None
+        # NOTE: This implementation remains stateless and loads the full volume
```
Comment (antonioterpin) [inline src/synthpix/data_sources/hdf5.py:31]: Should we write this in some documentation? Maybe preparing a data_loader.md

File: src/synthpix/data_sources/episodic.py [hunk 14]
```diff
@@ -0,0 +1,195 @@
...
+                f"source must be an instance of FileDataSource, got {type(source)}"
+            )
+
+        self.source = source
+        self.batch_size = batch_size
+        self.episode_length = episode_length
+        self._rng = np.random.default_rng(seed)
+
+        # 1. Discover Episodes from source.file_list
+        # We assume the source.file_list contains ALL valid files sorted/grouped by dir?
```
Comment (antonioterpin) [inline src/synthpix/data_sources/episodic.py]: We can validate it no? Or add a param, validate_structure. If true, we first validate and then run, else no. But at least we give the user a way to understand what is going on if errors are triggered.

Comment (antonioterpin) [reply src/synthpix/data_sources/episodic.py]: Probably I would put it with default true

Comment (antonioterpin) [reply src/synthpix/data_sources/episodic.py]: Actually, we can simply check it as we prepare the starts, no?

File: src/synthpix/data_sources/episodic.py [hunk 15]
```diff
@@ -0,0 +1,195 @@
...
+            List of tuples containing (file_path, chunk_id, timestep_in_episode).
+        """
+        # 1. Shuffle all possible episode starts
+        shuffled_starts = list(self._starts)  # copy
+        self._rng.shuffle(shuffled_starts)
+
+        # 2. Partition into chunks handling batch_size
+        num_chunks = len(shuffled_starts) // self.batch_size
+        # Note: We drop remainder episodes that don't fit into a full batch set.
+        # This is consistent with 'drop_remainder' batching logic often used in training.
```
Comment (antonioterpin) [inline src/synthpix/data_sources/episodic.py]: I do not think this is completely consistent with the dropping logic. I think we need to make sure we are dealing in a properly consistent way everywhere.. 

Example: For testing?

Num chunks should be one more if there is a remainder, and we can loop around when we build the chunk. Should be a easy fix.

Comment (antonioterpin) [reply src/synthpix/data_sources/episodic.py]: And of course we can set a flag for whether to do like this or not

**Review by antonioterpin — CHANGES_REQUESTED** [review]
Some comments

File: tests/test_data_generate.py [hunk 16]
```diff
@@ -85,27 +102,35 @@ def test_invalid_position_bounds(position_bounds):
...
-        ),
-        (
-            (-0.5, -0.5),
-            "seeding_density_range must be a tuple of two non-negative numbers.",
-        ),
-        ((1.0, 0.5), "seeding_density_range must be in the form \\(min, max\\)."),
-        ((0.5, 0.1), "seeding_density_range must be in the form \\(min, max\\)."),
-    ],
-)
+@pytest.mark.parametrize("seeding_density_range, expected_message",
```
Comment (antonioterpin) [inline tests/test_data_generate.py]: cant we have a nicer format? like break line after the first parenthesis of the parametrize

**Review by antonioterpin — CHANGES_REQUESTED** [review]
Only one comment

**Review by antonioterpin — APPROVED** [review]
